pkg/util/kernel/find_headers.go

```diff
@@ -0,0 +1,225 @@
+// +build linux
+
+package kernel
+
+import (
+	"bufio"
+	"bytes"
+	"fmt"
+	"io"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"regexp"
+	"strconv"
+
+	"github.com/DataDog/datadog-agent/pkg/metadata/host"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"github.com/mholt/archiver/v3"
+)
+
+const sysfsHeadersPath = "/sys/kernel/kheaders.tar.xz"
+const kernelModulesPath = "/lib/modules/%s/build"
+const debKernelModulesPath = "/lib/modules/%s/source"
+const cosKernelModulesPath = "/usr/src/linux-headers-%s"
+
+var versionCodeRegexp = regexp.MustCompile(`^#define[\t ]+LINUX_VERSION_CODE[\t ]+(\d+)$`)
+
+// HeaderFetchResult enumerates kernel header fetching success & failure modes
+type HeaderFetchResult int
+
+const (
+	// NotAttempted represents the case where runtime compilation fails prior to attempting to
+	// fetch kernel headers
+	NotAttempted HeaderFetchResult = iota
+	customHeadersFound
+	defaultHeadersFound
+	sysfsHeadersFound
+	downloadedHeadersFound
+	downloadSuccess
+	hostVersionErr
+	downloadFailure
+)
+
+// GetKernelHeaders attempts to find kernel headers on the host, and if they cannot be found it will attempt
+// to  download them to headerDownloadDir
+func GetKernelHeaders(headerDirs []string, headerDownloadDir string) ([]string, HeaderFetchResult, error) {
+	hv, hvErr := HostVersion()
+	if hvErr != nil {
+		return nil, hostVersionErr, fmt.Errorf("unable to determine host kernel version: %w", hvErr)
+	}
+
+	var err error
+	var dirs []string
+	if len(headerDirs) > 0 {
+		if err = validateHeaderDirs(hv, headerDirs); err == nil {
+			return headerDirs, customHeadersFound, nil
+		}
+		log.Debugf("unable to find configured kernel headers: %s", err)
+	} else {
+		dirs = getDefaultHeaderDirs()
+		if err = validateHeaderDirs(hv, dirs); err == nil {
+			return dirs, defaultHeadersFound, nil
+		}
+		log.Debugf("unable to find default kernel headers: %s", err)
+
+		// If no valid directories are found, attempt a fallback to extracting from `/sys/kernel/kheaders.tar.xz`
+		// which is enabled via the `kheaders` kernel module and the `CONFIG_KHEADERS` kernel config option.
+		// The `kheaders` module will be automatically added and removed if present and needed.
+		if dirs, err = getSysfsHeaderDirs(hv); err == nil {
+			return dirs, sysfsHeadersFound, nil
+		}
+		log.Debugf("unable to find system kernel headers: %s", err)
+	}
+
+	dirs = getDownloadedHeaderDirs(headerDownloadDir)
+	if err = validateHeaderDirs(hv, dirs); err == nil {
+		return dirs, downloadedHeadersFound, nil
+	}
+	log.Debugf("unable to find downloaded kernel headers: %s", err)
+
+	if err = downloadHeaders(headerDownloadDir); err == nil {
+		log.Infof("successfully downloaded kernel headers to %s", dirs)
```

Comment:
Do we need to validate the downloaded header dirs using `validateHeaderDirs`, or are we trusting that they are correct?

Reply:
I don't think it's likely that we'd download the wrong headers, but it doesn't hurt to validate them, so I'll add that